lib/astutils.cpp

```diff
@@ -3122,6 +3122,21 @@ struct FwdAnalysis::Result FwdAnalysis::checkRecursive(const Token *expr, const
                 // ({ .. })
                 if (hasGccCompoundStatement(parent->astParent()->astOperand2()))
                     return Result(Result::Type::BAILOUT);
+                auto isForLoop = [](const Token* tok) -> const Token* {
+                    for (int i = 0; i < 3 && tok; ++i)
+                        tok = tok->astParent();
+                    return tok && Token::simpleMatch(tok->previous(), "for (") ? tok : nullptr;
+                };
+                if (same) {
+                    const Token* forTok = isForLoop(tok);
+                    if (forTok) {
+                        Result resultFor = checkRecursive(expr, tok->astParent()->astOperand2(), forTok->link(), exprVarIds, local, true, depth);
```

Comment:
This seems to assume that `tok->astParent()` is `;`.

Once we have the top, we have functions like `getCondTok`, `getInitTok` and `getStepTok` to get the different statements used in the for loop. It seems like it would be better to use these functions.